include/lo2s/cupti/reader.hpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2016,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <lo2s/config.hpp>
+#include <lo2s/cupti/ringbuf.hpp>
+#include <lo2s/log.hpp>
+#include <lo2s/perf/time/converter.hpp>
+#include <lo2s/trace/trace.hpp>
+#include <lo2s/types.hpp>
+
+#include <chrono>
+#include <cstdlib>
+#include <string>
+
+extern "C"
+{
+#include <sys/timerfd.h>
+#include <unistd.h>
+}
+
+namespace lo2s
+{
+namespace cupti
+{
+
+class Reader
+{
+public:
+    Reader(trace::Trace& trace, Process process)
+    : process_(process), trace_(trace), time_converter_(perf::time::Converter::instance()),
+      ringbuf_reader_(std::to_string(process.as_pid_t()), 512)
+    {
+        struct itimerspec tspec;
+        memset(&tspec, 0, sizeof(struct itimerspec));
+        tspec.it_value.tv_nsec = 1;
+
+        tspec.it_interval.tv_sec =
+            std::chrono::duration_cast<std::chrono::seconds>(config().userspace_read_interval)
+                .count();
+
+        tspec.it_interval.tv_nsec =
+            (config().userspace_read_interval % std::chrono::seconds(1)).count();
+        timer_fd_ = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK);
+
+        timerfd_settime(timer_fd_, TFD_TIMER_ABSTIME, &tspec, NULL);
```

Comment:
This is the third reincarnation of building a tspec and using timerfd_settime. Refactor all these into a function taking a std::chrono::duration instead.